src/source/geojson_source.ts

```
@@ -172,6 +173,15 @@ class GeoJSONSource extends Evented implements Source {
         return this;
     }
 
+    updateData(diff: GeoJSONSourceDiff, getId: (feature: GeoJSON.Feature) => GeoJSONFeatureId) {
```

can we pass a "lookup" or "xpath" kind of parameter then to the worker to have the ability to search for ids with it somehow?
I know this is an annoying question, but I have data that I don't want to restructure just because this method is "too simple". I'm sure other users will have this issue as well...

---

So to effectively use partial updates with this source you're going to have to restructure how you call this anyways, so I don't think making an ID required is that big of a requirement.  There is also already precedent for features requiring IDs in the form of https://maplibre.org/maplibre-gl-js-docs/api/map/#map#setfeaturestate.

Making IDs required for partial updates now doesn't lock us into anything in the future, and we can always add support for extracting IDs from the properties in the future.

That said I did realize that there already exists some stuff that generates IDs:
https://maplibre.org/maplibre-gl-js-docs/style-spec/sources/#geojson-promoteId and https://maplibre.org/maplibre-gl-js-docs/style-spec/sources/#geojson-generateId

We probably need to throw an error if someone uses `updateData` with `generateId`.  Supporting `promoteId` seems reasonable, though that does mean this method loses type safety since we can't enforce the ID field being present statically.